Rotary Encoder: Don't call callbacks in the isr

mknjc opened the pull request:
Calling the on_clockwise and on_anticlockwise callbacks in an isr is not a good idea because they could do a lot of things which are not interrupt save (for example call delay())
Just accumulate the actions in a bitset and execute the callbacks in the correct order in loop

## Description:
I had a lot of crashes with a simple monochromatic light, a rotary encoder which sets the relative brightness of the light and a connected Home Assistant instance. Fast rotations overflow the tcp send buffer which causes rescheduling attempts which can't work in interrupts.
This is one possible way I came up and I had no problems after this.
Its not optimal because if you cause more than the 128 events between two loop calls (quite easy with debug logging on and a few connected clients) some events gets dropped but I think in reality its quite unrealistic to have more than 128 events in 15ms

**Related issue (if applicable):** nothing found

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** I don't think mentioning in the docs is necessary?

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder). < I don't see how this could currently be tested

mknjc committed:
Rotary Encoder: Don't call callbacks in the isr

Calling the on_clockwise and on_anticlockwise callbacks in an isr is not a good idea because they could do a lot of things which are not interrupt save (for example call delay())
Just accumulate the actions in a bitset and execute the callbacks in the correct order in loop

badbadc0ffee commented:
To further reduce the risk of loosing events, one could consider a short (signed) byte vector instead of the bitvector. Unless the direction changes or the event number in one direction reaches 127, you could accumulate them all in one byte.
While it is already *quite unrealistic* that more than 128 events would occur between two `loop()` calls, it is probably *even more unrealistic* to have a lot of direction changes in such a short period of time.

jesserockz commented:
I quite like what has been done here. Thank you.

One thing would be also moving the callbacks out of the store and into the sensor as they are not needed in the store now.
```cpp
  CallbackManager<void()> on_clockwise_callback_;
  CallbackManager<void()> on_anticlockwise_callback_;
```

I think realistically with normal use we can expect the number of steps to be less than 128 in a single `loop()` so this would be fine.

mknjc commented:
> To further reduce the risk of loosing events, one could consider a short (signed) byte vector instead of the bitvector. Unless the direction changes or the event number in one direction reaches 127, you could accumulate them all in one byte.
> While it is already _quite unrealistic_ that more than 128 events would occur between two `loop()` calls, it is probably _even more unrealistic_ to have a lot of direction changes in such a short period of time.

I really like the idea, we could use 4 or 8 bytes instead of the current 17 bytes and store up to 1016 events with this.

mknjc committed:
Rotary Encoder: Move rotation callbacks into the Component

mknjc committed:
Rotary Encoder: store rotation events in a eventlist

This could hold up to 8 times 127 rotations in one direction or up to 8 direction changes between calls to loop
